staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -298,14 +298,10 @@ func TestHammerController(t *testing.T) {
 	time.Sleep(100 * time.Millisecond)
 	cancel()
 
-	// Before we permanently lock this mutex, we have to be sure
-	// that the controller has stopped running. At this point,
-	// all goroutines should have stopped. Leak checking is
-	// done by TestMain.
+	// Wait until the controller has fully stopped running.
+	// At this point, all goroutines should have stopped, and
+	// leak checking is done by TestMain.
 	controllerWG.Wait()
-
-	outputSetLock.Lock()
```

Comment:
Removing this `Lock` has the effect that the test no longer works as originally intended. It was already there before I added `WaitGroup` waiting.

I didn't write it, so I can't say whether the controller logic indeed ensures that all goroutines have stopped once `controller.RunWithContext` returns.

Reply:
Given that the original test didn't actually check for leaked goroutines (see the TODO that I replaced with the comment about leak checking in main with goleak), it is quite possible that there is a flaw in the controller logic that hasn't been detected earlier.

Reply:
so , most possible the test is right and there is some subtle leak in the controller

Reply:
Thanks for catching that, and sorry again for the confusion. I'll close this PR and open a new issue to look into the controller leak that's showing up.

Reply:
anyway, was a good exercise, so we have more clear this test, I think it was a positive outcome overall